src/de.rs

```diff
@@ -262,80 +268,19 @@ impl<'de, 'a, 'b> serde::Deserializer<'de> for &'a mut Deserializer<'b> {
     where
         V: Visitor<'de>,
     {
-        let marker = self.read_marker()?;
-        self.dispatch_based_on_marker(marker, visitor)
-    }
-
-    fn deserialize_bool<V>(self, visitor: V) -> Result<<V as Visitor<'de>>::Value>
-    where
-        V: Visitor<'de>,
-    {
-        self.read_expected_marker(MARKER_SINGLE_BOOL)?;
-        self.dispatch_based_on_marker(MARKER_SINGLE_BOOL, visitor)
-    }
-
-    fn deserialize_i8<V>(self, visitor: V) -> Result<<V as Visitor<'de>>::Value>
-    where
-        V: Visitor<'de>,
-    {
-        self.read_expected_marker(MARKER_SINGLE_I8)?;
-        self.dispatch_based_on_marker(MARKER_SINGLE_I8, visitor)
-    }
-
-    fn deserialize_i16<V>(self, visitor: V) -> Result<<V as Visitor<'de>>::Value>
-    where
-        V: Visitor<'de>,
-    {
-        self.read_expected_marker(MARKER_SINGLE_I16)?;
-        self.dispatch_based_on_marker(MARKER_SINGLE_I16, visitor)
-    }
-
-    fn deserialize_i32<V>(self, visitor: V) -> Result<<V as Visitor<'de>>::Value>
-    where
-        V: Visitor<'de>,
-    {
-        self.read_expected_marker(MARKER_SINGLE_I32)?;
-        self.dispatch_based_on_marker(MARKER_SINGLE_I32, visitor)
-    }
-
-    fn deserialize_i64<V>(self, visitor: V) -> Result<<V as Visitor<'de>>::Value>
-    where
-        V: Visitor<'de>,
-    {
-        self.read_expected_marker(MARKER_SINGLE_I64)?;
-        self.dispatch_based_on_marker(MARKER_SINGLE_I64, visitor)
-    }
-
-    fn deserialize_u8<V>(self, visitor: V) -> Result<<V as Visitor<'de>>::Value>
-    where
-        V: Visitor<'de>,
-    {
-        self.read_expected_marker(MARKER_SINGLE_U8)?;
-        self.dispatch_based_on_marker(MARKER_SINGLE_U8, visitor)
-    }
-
-    fn deserialize_u16<V>(self, visitor: V) -> Result<<V as Visitor<'de>>::Value>
-    where
-        V: Visitor<'de>,
-    {
-        self.read_expected_marker(MARKER_SINGLE_U16)?;
-        self.dispatch_based_on_marker(MARKER_SINGLE_U16, visitor)
-    }
-
-    fn deserialize_u32<V>(self, visitor: V) -> Result<<V as Visitor<'de>>::Value>
-    where
-        V: Visitor<'de>,
-    {
-        self.read_expected_marker(MARKER_SINGLE_U32)?;
-        self.dispatch_based_on_marker(MARKER_SINGLE_U32, visitor)
+        if !self.read_header {
+            self.read_header = true;
+            visitor.visit_map(MapAccess::with_varint_encoded_fields(self)?)
```

Comment:
This change seems to be orthogonal, correct? Can we have that in a separate PR please?